### PR TITLE
[V1][Spec Decode] Enable spec decode for top-p & top-k sampling

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR is to enable spec decoding for requests with top-p & top-k sampling.
It is implemented using `apply_top_k_top_p` to mask the logits of the target model.
While this is more expensive than FlashInfer's sorting-free sampling, I think it's good for the first step.

### CODE REVIEW

File: vllm/v1/sample/rejection_sampler.py [hunk 1]
```diff
@@ -245,25 +246,85 @@ def compute_probs(
...
-        1,  # replace_to
-        MAX_NUM_TOKENS=MAX_SPEC_LEN,
-        num_warps=1,
+        num_tokens,
+        replace_from=GREEDY_TEMPERATURE,
+        replace_to=1,
     )
-    output_prob = compiled_softmax(logits, expanded_temperature)
+    # TODO(woosuk): Consider using in-place op to reduce memory usage.
+    logits = logits / temperature.unsqueeze(-1)
```
Comment (houseroad) [inline vllm/v1/sample/rejection_sampler.py:257]: inplace: 

logits.div_(temparture.unsqueeze(-1))

**Review by houseroad — COMMENTED** [review]
Looks good. Wondering how did we test it?

File: vllm/v1/sample/rejection_sampler.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/sample/rejection_sampler.py:257]: I'm a bit hesitant because if we do it, we need to take care of the side effect. Basically making sure no other op after rejection sampler uses the (updated) logits.

**Comment by WoosukKwon** [comment]
> Looks good. Wondering how did we test it?

Good point. I just wanted to get some initial feedback before adding tests. Will update the PR.

**Comment by WoosukKwon** [comment]
@houseroad @LiuXiaoxuanPKU I've added the tests, and they're passing locally. Could you please review? I'd appreciate including this PR in v0.8.2 if possible.

File: tests/v1/sample/test_rejection_sampler.py [hunk 2]
```diff
@@ -462,3 +464,136 @@ def estimate_rejection_sampling_pdf(
...
+@pytest.mark.parametrize("top_p", [0.5, 0.9, 0.99])
+def test_top_p(rejection_sampler, top_p):
+    """Test rejection sampling with top-p sampling"""
+    vocab_size = 100
+    batch_size = 100
+    num_draft_tokens = 3
+    num_tokens = batch_size * num_draft_tokens
+
+    # Randomly create unmasked indices.
+    num_top_p_tokens = int(vocab_size * top_p)
```
Comment (LiuXiaoxuanPKU) [inline tests/v1/sample/test_rejection_sampler.py:575]: A bit confused by the definition of `top_p` sampling, should it be ' restricting the sampling to the set of most probable tokens with cumulative probability more than p'? Instead of sampling a fixed percentage of tokens like here.

**Review by LiuXiaoxuanPKU — COMMENTED** [review]
LGTM, just minor QQ about the test

File: tests/v1/sample/test_rejection_sampler.py (hunk 2 above)
Comment (WoosukKwon) [reply tests/v1/sample/test_rejection_sampler.py]: @LiuXiaoxuanPKU Good catch. It only makes sense when `int(vocab_size * top_p)` tokens all have equal high logits (e.g., 100) while the others have -100. But definitely this is not general enough.

I've updated it to test top-p more precisely. Could you please take another look?

**Review by njhill — APPROVED** [review]
LGTM

File: vllm/v1/sample/rejection_sampler.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/sample/rejection_sampler.py:257]: @njhill Good point. After second thought, I actually planned to do it in a followup PR. Let's do it in the next PR!

File: vllm/v1/sample/rejection_sampler.py [hunk 3]
```diff
@@ -245,25 +246,85 @@ def compute_probs(
...
+    Returns:
+        expanded_x: [num_tokens] tensor.
+    """
+    batch_size = x.shape[0]
+    assert cu_num_tokens.shape[0] == batch_size
+    expanded_x = torch.empty(
+        (num_tokens, ),
+        dtype=x.dtype,
+        device=x.device,
+    )
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/sample/rejection_sampler.py]: good catch. updated.
